source/adapters/opencl/command_buffer.cpp

```diff
@@ -180,19 +277,57 @@ ur_result_t UR_APICALL urCommandBufferAppendMembufferReadRectExp(
         *pSyncPointWaitList,
     [[maybe_unused]] ur_exp_command_buffer_sync_point_t *pSyncPoint) {
 
-  cl_adapter::die("Experimental Command-buffer feature is not "
+  cl_adapter::die("Experimental Command-buffer entry point is not "
                   "implemented for OpenCL adapter.");
   return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
 }
 
+UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferAppendMembufferFillExp(
+    ur_exp_command_buffer_handle_t hCommandBuffer, ur_mem_handle_t hBuffer,
+    const void *pPattern, size_t patternSize, size_t offset, size_t size,
+    uint32_t numSyncPointsInWaitList,
+    const ur_exp_command_buffer_sync_point_t *pSyncPointWaitList,
+    ur_exp_command_buffer_sync_point_t *pSyncPoint) {
+
+  cl_context CLContext = cl_adapter::cast<cl_context>(hCommandBuffer->hContext);
+  cl_ext::clCommandFillBufferKHR_fn F = nullptr;
+  cl_int Res = cl_ext::getExtFuncFromContext<decltype(F)>(
+      CLContext, cl_ext::ExtFuncPtrCache->clCommandFillBufferKHRCache,
+      cl_ext::CommandFillBufferName, &F);
+
+  if (!F || Res != CL_SUCCESS)
+    return UR_RESULT_ERROR_INVALID_OPERATION;
+
+  CL_RETURN_ON_FAILURE(
+      F(hCommandBuffer->CLCommandBuffer,
+        cl_adapter::cast<cl_command_queue>(hCommandBuffer->hInternalQueue),
+        cl_adapter::cast<cl_mem>(hBuffer), pPattern, patternSize, offset, size,
+        numSyncPointsInWaitList, pSyncPointWaitList, pSyncPoint, nullptr));
+
+  return UR_RESULT_SUCCESS;
+}
+
 UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
-    [[maybe_unused]] ur_exp_command_buffer_handle_t hCommandBuffer,
-    [[maybe_unused]] ur_queue_handle_t hQueue,
-    [[maybe_unused]] uint32_t numEventsInWaitList,
-    [[maybe_unused]] const ur_event_handle_t *phEventWaitList,
-    [[maybe_unused]] ur_event_handle_t *phEvent) {
+    ur_exp_command_buffer_handle_t hCommandBuffer, ur_queue_handle_t hQueue,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    ur_event_handle_t *phEvent) {
 
-  cl_adapter::die("Experimental Command-buffer feature is not "
-                  "implemented for OpenCL adapter.");
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+  cl_context CLContext = cl_adapter::cast<cl_context>(hCommandBuffer->hContext);
+  cl_ext::clEnqueueCommandBufferKHR_fn F = nullptr;
+  cl_int Res = cl_ext::getExtFuncFromContext<decltype(F)>(
+      CLContext, cl_ext::ExtFuncPtrCache->clEnqueueCommandBufferKHRCache,
+      cl_ext::EnqueueCommandBufferName, &F);
+
+  if (!F || Res != CL_SUCCESS)
+    return UR_RESULT_ERROR_INVALID_OPERATION;
+
+  uint32_t numQueues = (hQueue) ? 1 : 0;
```

Comment:
Why do we need this check? [urCommandBufferEnqueueExp](https://oneapi-src.github.io/unified-runtime/core/api.html#urcommandbufferenqueueexp) specifies that if `hQueue` is NULL then `UR_RESULT_ERROR_INVALID_NULL_HANDLE` should be returned

Reply:
I added this from reading the below from the [OpenCL spec](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#_add_new_section_section_5_x_2_enqueuing_a_command_buffer).
```
queues A pointer to an ordered list of command-queues [compatible](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#compatible) with the command-queues used on recording. queues can be NULL in which case the default command-queues used on command-buffer creation are used and num_queues must be 0.
```

Reply:
Sure, but that OpenCL behaviour isn't a valid implementation of this UR entry-point. For example, if we had a CTS test that tested calling `urCommandBufferEnqueueExp` with a null `hQueue` parameter, then the expected behaviour is to return an error `UR_RESULT_ERROR_INVALID_NULL_HANDLE `, not enqueue the command-buffer.

Reply:
Apologies that's my misunderstanding. I wasn't aware of how the validation layer works before but have now found out those checks specified in the spec are done [in that layer](https://github.com/oneapi-src/unified-runtime/blob/main/source/loader/layers/validation/ur_valddi.cpp#L7434). I've removed the `uint32_t numQueues = (hQueue) ? 1 : 0;`. Cheers.